services/src/test/java/org/keycloak/compatibility/KeycloakCompatibilityMetadataProviderTest.java

```diff
@@ -0,0 +1,72 @@
+package org.keycloak.compatibility;
+
+import org.junit.Test;
+import org.keycloak.common.Profile;
+import org.keycloak.common.profile.ProfileConfigResolver;
+
+import java.util.Map;
+
+import static org.junit.Assert.assertEquals;
+import static org.keycloak.compatibility.KeycloakCompatibilityMetadataProvider.VERSION_KEY;
+
+public class KeycloakCompatibilityMetadataProviderTest {
+
+    @Test
+    public void testMicroVersionUpgradeWorksWithRollingUpdateV2() {
+        // Enable V2 feature
+        Profile.configure(new ProfileConfigResolver() {
+            @Override
+            public Profile.ProfileName getProfileName() {
+                return null;
+            }
+
+            @Override
+            public FeatureConfig getFeatureConfig(String feature) {
+                return Profile.Feature.ROLLING_UPDATES_V2.getVersionedKey().equals(feature) ? FeatureConfig.ENABLED : FeatureConfig.UNCONFIGURED;
+            }
+        });
+
+        // Make compatibility provider return hardcoded version as we are not able to test this in integration tests with micro versions equal to 0
+        KeycloakCompatibilityMetadataProvider compatibilityProvider = new KeycloakCompatibilityMetadataProvider("999.999.999-Final");
+
+        // Test compatible
+        assertCompatibility(CompatibilityResult.ExitCode.ROLLING, compatibilityProvider.isCompatible(Map.of(VERSION_KEY, "999.999.999-Final")));
+        assertCompatibility(CompatibilityResult.ExitCode.ROLLING, compatibilityProvider.isCompatible(Map.of(VERSION_KEY, "999.999.998-Final")));
+        assertCompatibility(CompatibilityResult.ExitCode.ROLLING, compatibilityProvider.isCompatible(Map.of(VERSION_KEY, "999.999.999-Final1")));
+        assertCompatibility(CompatibilityResult.ExitCode.ROLLING, compatibilityProvider.isCompatible(Map.of(VERSION_KEY, "999.999.1-Final")));
+
+        // Test incompatible
+        assertCompatibility(CompatibilityResult.ExitCode.RECREATE, compatibilityProvider.isCompatible(Map.of(VERSION_KEY, "999.999.1000-Final")));
```

Comment:
why is this incompatible? `999.999.999-Final` to `999.999.1000-Final` should be compatible, right!?

Reply:
This is testing `999.999.1000-Final` to `999.999.999-Final`. I don't think we should do rollback in rolling way. WDYT?

Reply:
Oh, I misread the test then. I'm fine preventing a rollback.